### PR TITLE
Rework interface

### PR DESCRIPTION
**Pull request by pierre-wilmot-fetch-ai**

### CODE REVIEW

File: libs/ml/examples/basic/main.cpp [hunk 1]
```diff
@@ -57,11 +57,10 @@ int main(int ac, char **av)
 
   CrossEntropy<ArrayType> criterion;
 
-  std::pair<size_t, std::shared_ptr<ArrayType>> input;
-  std::shared_ptr<ArrayType>                    gt =
-      std::make_shared<ArrayType>(std::vector<typename ArrayType::SizeType>({1, 10}));
+  std::pair<size_t, ArrayType> input;
```
Comment (baykaner) [inline libs/ml/examples/basic/main.cpp]: std::size_t please

File: libs/ml/include/ml/ops/activations/softmax.hpp [hunk 2]
```diff
@@ -35,40 +35,31 @@ class Softmax : public fetch::ml::Ops<T>
...
-    for (std::size_t i(0); i < inputs[0]->size(); ++i)
-    {
-      sum += errorSignal->At(i);
-    }
-    for (std::size_t i(0); i < inputs[0]->size(); ++i)
-    {
-      errorSignal->At(i) -= (t->At(i) * sum);
-    }
-    return {errorSignal};
+    ArrayType returnSignal = errorSignal.Clone();
```
Comment (baykaner) [inline libs/ml/include/ml/ops/activations/softmax.hpp:56]: nice - this is all much cleaner

File: libs/ml/include/ml/ops/ops.hpp [hunk 3]
```diff
@@ -31,12 +31,13 @@ class Ops
...
-  virtual std::vector<ArrayPtrType> Backward(std::vector<ArrayPtrType> const &inputs,
-                                             ArrayPtrType                     error)                     = 0;
+  virtual ArrayType Forward(std::vector<std::reference_wrapper<ArrayType const>> const &inputs) = 0;
+  virtual std::vector<ArrayType> Backward(
+      std::vector<std::reference_wrapper<ArrayType const>> const &inputs,
+      ArrayType const &                                           errorSignal) = 0;
 
 protected:
-  ArrayPtrType output_;
+  ArrayPtrType output_;  // TODO : remove
```
Comment (baykaner) [inline libs/ml/include/ml/ops/ops.hpp]: please create and issues and attach issue numbers to TODOs

File: libs/ml/tests/ml/ops/activations/softmax.cpp [hunk 4]
```diff
@@ -33,43 +33,43 @@ TYPED_TEST_CASE(SoftmaxTest, MyTypes);
 
 TYPED_TEST(SoftmaxTest, forward_test)
 {
-  std::shared_ptr<TypeParam> data = std::make_shared<TypeParam>(8);
-  std::shared_ptr<TypeParam> gt   = std::make_shared<TypeParam>(8);
-  std::vector<double>        dataInput({1, -2, 3, -4, 5, -6, 7, -8});
-  std::vector<double>        gtInput({2.1437e-03, 1.0673e-04, 1.5840e-02, 1.4444e-05, 1.1704e-01,
+  TypeParam           data(8);
```
Comment (baykaner) [inline libs/ml/tests/ml/ops/activations/softmax.cpp:36]: i love how much cleaner tests are now 👍

File: libs/ml/tests/ml/ops/matrix_multiply.cpp [hunk 5]
```diff
@@ -45,37 +45,35 @@ TYPED_TEST(MatrixMultiplyTest, forward_test)
...
 }
 
 TYPED_TEST(MatrixMultiplyTest, backward_test)
 {
-  std::shared_ptr<TypeParam> a  = std::make_shared<TypeParam>(std::vector<std::uint64_t>({1, 5}));
-  std::shared_ptr<TypeParam> b  = std::make_shared<TypeParam>(std::vector<std::uint64_t>({5, 4}));
-  std::shared_ptr<TypeParam> e  = std::make_shared<TypeParam>(std::vector<std::uint64_t>({1, 4}));
-  std::shared_ptr<TypeParam> ig = std::make_shared<TypeParam>(std::vector<std::uint64_t>({1, 5}));
-  std::shared_ptr<TypeParam> wg = std::make_shared<TypeParam>(std::vector<std::uint64_t>({5, 4}));
+  TypeParam a(std::vector<std::uint64_t>({1, 5}));
```
Comment (baykaner) [inline libs/ml/tests/ml/ops/matrix_multiply.cpp:72]: not very meaningful variable names in this test; do the letters actually mean something?

**Review by baykaner — DISMISSED** [review]
Please raise an issue and attach an issue number to the TODO about removing ouput_ from Ops.

File: libs/ml/tests/ml/ops/matrix_multiply.cpp (hunk 5 above)
Comment (pierre-wilmot-fetch-ai) [reply libs/ml/tests/ml/ops/matrix_multiply.cpp:72]: A, B -- The two matrices being multiplied
E - Error
ig -- Input gradient
wg -- Weight gradient

I'll try to go with something more explicit, using a name based on weigthts doesn't makes sense here

**Review by baykaner — APPROVED** [review]

**Pull request merged**